cmake/Toolchain-gcc-clang.cmake

```diff
@@ -0,0 +1,216 @@
+        set(SANITIZER address)
+    endif()
+    list(PREPEND CMAKE_REQUIRED_LIBRARIES -fsanitize=${SANITIZER})
+    check_cxx_compiler_flag("-fsanitize=${SANITIZER}" ASAN_SUPPORT_FLAG)
+    if(${ASAN_SUPPORT_FLAG})
+        list(PREPEND MY_C_DBG_FLAGS -fsanitize=${SANITIZER})
+    else()
+        message(FATAL_ERROR "asan not available to compiler.")
+    endif()
+endif()
+
+if(ENABLE_SSP AND CMAKE_BUILD_TYPE STREQUAL Debug)
+    if(WIN32)
+        set(SSP_STATIC ON)
+    endif()
+
+    find_package(SSP)
+
+    if(SSP_LIBRARY)
+        list(APPEND MY_C_FLAGS        -D_FORTIFY_SOURCE=2)
+        list(APPEND MY_C_LINKER_FLAGS ${SSP_LIBRARY})
+    endif()
+endif()
+
+if(NOT (WIN32 OR X86_32)) # inline asm is not allowed with -fPIC
+    set(MY_C_FLAGS ${MY_C_FLAGS} -fPIC)
+endif()
+
+# check if ssp flags are supported for this version of gcc
+if(CMAKE_COMPILER_IS_GNUCXX)
+    if(ENABLE_SSP)
+        check_cxx_compiler_flag(-fstack-protector-strong  F_STACK_PROTECTOR_STRONG_FLAG)
+
+        if(F_STACK_PROTECTOR_STRONG_FLAG)
+            set(MY_C_FLAGS ${MY_C_FLAGS} -fstack-protector-strong)
+
+            check_cxx_compiler_flag("--param ssp-buffer-size=4" SSP_BUFFER_SIZE_FLAG)
+
+            if(SSP_BUFFER_SIZE_FLAG)
+                # we do not add it to MY_C_FLAGS because this breaks things like CMAKE_REQUIRED_LIBRARIES
+                # which misinterpret compiler flags without leading dashes
+                add_compile_options(--param ssp-buffer-size=4)
+            endif()
+        endif()
+    endif()
+
+    set(MY_C_FLAGS ${MY_C_FLAGS} -fopenmp)
+endif()
+
+if(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
+    unset(COMPILER_COLOR_DIAGNOSTICS)
+    check_cxx_compiler_flag(-fdiagnostics-color=always COMPILER_COLOR_DIAGNOSTICS)
+    if(COMPILER_COLOR_DIAGNOSTICS)
+        add_compile_options(-fdiagnostics-color=always)
+    else()
+        unset(COMPILER_COLOR_DIAGNOSTICS)
+        check_cxx_compiler_flag(-fdiagnostics-color COMPILER_COLOR_DIAGNOSTICS)
+        if(COMPILER_COLOR_DIAGNOSTICS)
+            add_compile_options(-fdiagnostics-color)
+        endif()
+    endif()
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL Clang)
+    unset(COMPILER_COLOR_DIAGNOSTICS)
+    check_cxx_compiler_flag(-fcolor-diagnostics COMPILER_COLOR_DIAGNOSTICS)
+    if(COMPILER_COLOR_DIAGNOSTICS)
+        add_compile_options(-fcolor-diagnostics)
+    endif()
+endif()
+
+if(MINGW)
+    set(MY_C_FLAGS ${MY_C_FLAGS} -static-libgcc -static-libstdc++)
+endif()
+
+if(CMAKE_BUILD_TYPE STREQUAL Debug)
+    set(MY_C_FLAGS   ${MY_C_FLAGS}   ${MY_C_DBG_FLAGS} -Wall -Wextra)
+else()
+    set(MY_C_FLAGS   ${MY_C_FLAGS}   ${MY_C_OPT_FLAGS} -Wno-error)
+endif()
+
+# for some reason this is necessary
+if(CMAKE_SYSTEM_NAME STREQUAL FreeBSD)
+    set(MY_C_FLAGS -I/usr/local/include ${MY_C_FLAGS})
+endif()
+
+foreach(C_COMPILE_FLAG ${MY_C_FLAGS})
+    add_compile_options(${C_COMPILE_FLAG})
+endforeach()
+
+include(CheckCXXCompilerFlag)
+
+check_cxx_compiler_flag(-std=gnu++17 GNUPP17_FLAG)
```

Comment:
I'm not sure if we should do this anymore because we're now setting the C++ standard with cmake.

Reply:
I removed it. In fact, I rewrote the entire file.